include/ipc/ipc_service.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+#define ZEPHYR_INCLUDE_IPC_SERVICE_IPC_SERVICE_H_
+
+#include <stdio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief IPC Service API
+ * @defgroup ipc_service_api IPC service APIs
+ * @{
+ */
+
+/** @brief Event callback structure
+ *  It is registered during endpoint registration.
+ *  This structure is packed into the endpoint configuration.
+ */
+struct ipc_service_cb {
+	/** @brief Bind was successful
+	 *  @param priv Private user data.
+	 */
+	int (*bound)(void *priv);
+
+	/** @brief New packet arrived.
+	 *  @param data Pointer to data buffer.
+	 *  @param len Length of @a data.
+	 *  @param priv Private user data.
+	 */
+	int (*received)(void *data, size_t len, void *priv);
+
+	/** @brief An error occurred
+	 *  @param message Error message.
+	 *  @param priv Private user data.
+	 */
+	int (*error)(const char *message, void *priv);
+};
```

Comment:
In this implementation, the callback returns a value. This was done with the idea of passing the value to the backend.
Similar to the callback implementation from RPMsg:
`typedef int (*rpmsg_ept_cb)(struct rpmsg_endpoint *ept, void *data,
			    size_t len, uint32_t src, void *priv);`

It seems to me that it will be necessary to add information to the description what value should be returned.